swift/Arrow/Sources/Arrow/ArrowReader.swift

```diff
@@ -54,49 +86,104 @@ public class ArrowReader {
+                                         length: nullLength, messageOffset: loadInfo.messageOffset)
+        var children = [ArrowData]()
+        for index in 0..<field.childrenCount {
+            let childField = field.children(at: index)!
+            switch loadField(loadInfo, field: childField) {
+            case .success(let holder):
+                children.append(holder.array.arrowData)
+            case .failure(let error):
+                return .failure(error)
+            }
         }
+
+        return makeArrayHolder(field, buffers: [arrowNullBuffer],
+                               nullCount: UInt(node.nullCount), children: children,
+                               rbLength: UInt(loadInfo.batchData.recordBatch.length))
     }
 
-    private func loadVariableData(_ loadInfo: DataLoadInfo) -> Result<ArrowArrayHolder, ArrowError> {
-        let node = loadInfo.recordBatch.nodes(at: loadInfo.nodeIndex)!
-        do {
-            let nullLength = UInt(ceil(Double(node.length) / 8))
-            try validateBufferIndex(loadInfo.recordBatch, index: loadInfo.bufferIndex)
-            let nullBuffer = loadInfo.recordBatch.buffers(at: loadInfo.bufferIndex)!
-            let arrowNullBuffer = makeBuffer(nullBuffer, fileData: loadInfo.fileData,
-                                             length: nullLength, messageOffset: loadInfo.messageOffset)
-            try validateBufferIndex(loadInfo.recordBatch, index: loadInfo.bufferIndex + 1)
-            let offsetBuffer = loadInfo.recordBatch.buffers(at: loadInfo.bufferIndex + 1)!
-            let arrowOffsetBuffer = makeBuffer(offsetBuffer, fileData: loadInfo.fileData,
-                                               length: UInt(node.length), messageOffset: loadInfo.messageOffset)
-            try validateBufferIndex(loadInfo.recordBatch, index: loadInfo.bufferIndex + 2)
-            let valueBuffer = loadInfo.recordBatch.buffers(at: loadInfo.bufferIndex + 2)!
-            let arrowValueBuffer = makeBuffer(valueBuffer, fileData: loadInfo.fileData,
-                                              length: UInt(node.length), messageOffset: loadInfo.messageOffset)
-            return makeArrayHolder(loadInfo.field, buffers: [arrowNullBuffer, arrowOffsetBuffer, arrowValueBuffer],
-                                   nullCount: UInt(node.nullCount))
-        } catch let error as ArrowError {
-            return .failure(error)
-        } catch {
-            return .failure(.unknownError("\(error)"))
+    private func loadPrimitiveData(
+        _ loadInfo: DataLoadInfo,
+        field: org_apache_arrow_flatbuf_Field)
+    -> Result<ArrowArrayHolder, ArrowError> {
+        guard let node = loadInfo.batchData.nextNode() else {
+            return .failure(.invalid("Node not found"))
+        }
+
+        guard let nullBuffer = loadInfo.batchData.nextBuffer() else {
+            return .failure(.invalid("Buffer not found"))
```

Comment:
How about adding more information which buffer type isn't found?

```suggestion
            return .failure(.invalid("Null buffer not found"))
```